### ISSUE TITLE
Use a persistent datastructure for fields

### ISSUE THREAD

**Issue by sdboyer** [sirupsen/logrus#234]
Hiya -

I'm curious if you've considered using a persistent datastructure, e.g. github.com/mndrix/ps, for `Fields` in place of `map[string]interface{}`. The thought occurred to me when I looked at `Entry.WithFields()`, and I noticed it constructs a new `Entry` by copying out the fields in the current `Entry`, then doing the same with the map from the parameter.

A persistent structure would obviate the need for the copy, at least. But more substantively, it _might_ result in fewer overall allocations for logrus...though I'm not familiar enough with the compiler to really be sure about that off the top of my head. Some basic escape analysis (`gcflags=-m`) leaves me unsure. If it does reduce allocs - particularly allocs that happen regardless of the set logging level - that could make logrus' footprint even lighter.

So basically, I thought I'd open an issue to see if you had any interest in exploring this. If you think it'd be worth at least seeing the results, I'd be happy to code something up. :computer: 

(Of course, #207 is pretty much a prerequisite for this)

**Comment by sdboyer** [comment]
Oh, also, a `sync.Pool` could reduce allocs for some cases a bit, as well.

**Comment by Sirupsen** [comment]
Let's tackle #207 and take a look at this. Definitely a reasonable thing to experiment with.

**Comment by Sirupsen** [comment]
I'm tagging this with 1.0 as it's something to experiment with here.

**Comment by markphelps** [comment]
Closing per the project being put into [maintenance mode](https://github.com/sirupsen/logrus/commit/4ecd9a62bd57b6533e0d87ae23a6b10adc32b5ce)
